src/kvstore/wal/FileBasedWal.cpp

```diff
@@ -450,69 +405,68 @@ bool FileBasedWal::appendLogInternal(BufferPtr& buffer,
         return false;
     }
 
-    if (buffer &&
-        (buffer->size() +
-         msg.size() +
-         sizeof(ClusterID) +
-         sizeof(TermID) +
-         sizeof(LogID) > maxBufferSize_)) {
-        // Freeze the current buffer
-        if (buffer->freeze()) {
-            flusher_->flushBuffer(shared_from_this(), buffer);
-        }
-        buffer.reset();
-    }
+    // Write to the WAL file first
+    std::string strBuf;
+    strBuf.reserve(sizeof(LogID)
+                    + sizeof(TermID)
+                    + sizeof(ClusterID)
+                    + msg.size()
+                    + 2 * sizeof(int32_t));
+    strBuf.append(reinterpret_cast<char*>(&id), sizeof(LogID));
+    strBuf.append(reinterpret_cast<char*>(&term), sizeof(TermID));
+    int32_t len = msg.size();
+    strBuf.append(reinterpret_cast<char*>(&len), sizeof(int32_t));
+    strBuf.append(reinterpret_cast<char*>(&cluster), sizeof(ClusterID));
+    strBuf.append(reinterpret_cast<const char*>(msg.data()), msg.size());
+    strBuf.append(reinterpret_cast<char*>(&len), sizeof(int32_t));
+
+    // Prepare the WAL file if it's not opened
+    if (currFd_ < 0) {
+        prepareNewFile(id);
+    } else if (currInfo_->size() + strBuf.size() > maxFileSize_) {
+        // Need to roll over
+        closeCurrFile();
 
-    // Create a new buffer if needed
-    if (!buffer) {
-        std::unique_lock<std::mutex> guard(buffersMutex_);
-        buffer = createNewBuffer(id, guard);
+        std::lock_guard<std::mutex> g(walFilesMutex_);
+        prepareNewFile(id);
     }
 
-    DCHECK_EQ(
-        id,
-        static_cast<int64_t>(buffer->firstLogId() + buffer->numLogs()));
-    buffer->push(term, cluster, std::move(msg));
+    ssize_t bytesWritten = write(currFd_, strBuf.data(), strBuf.size());
```

Comment:
We could use _gather output_ to do this, aka. `pwritev`, which could save us one copy operation, which means a lot here.

Of course we could do it in a separate PR. Please let me take care of it.

Reply:
Ok, I'll leave it out of this PR